Hotfix 1.9.5 and release notes (#5588)
dapr/dapr: docs/release_notes/v1.9.5.md

# Dapr 1.9.5

> - Fixes an issue in the CosmosDB state store component when performing transaction request with ETags

Reviewer: ```suggestion
- Fixes a panic in the Azure Cosmos DB state store component when performing transaction request with ETags
```

- Fixes component initialization failure when the built-in Kubernetes secret store is disabled

> ## Fixes an issue in the CosmosDB state store component when performing transaction request with ETags

Reviewer: ```suggestion
## Fixes a panic in the Azure Cosmos DB state store component when performing transaction request with ETags
```

### Problem

> When setting any ETag in a transaction request inside CosmosDB, dapr exits due to a nil pointer dereference.

Reviewer: ```suggestion
When setting any ETag in a transaction request with the Azure Cosmos DB state store, Dapr panics with a nil pointer dereference.
```

### Impact

> This issue impacts all users of CosmosDB state store.

Reviewer: ```suggestion
This issue impacts all users of the Azure Cosmos DB state store on Dapr 1.9.0-1.9.4.
```

### Root cause

> When setting an ETag on a transaction request in CosmosDB state store, a nil pointer was being dereferenced. 

Reviewer: ```suggestion
When setting an ETag on a transaction request in the Azure Cosmos DB state store, a bug causes Dapr to panic.
```

### Solution

> We fixed a bug that caused Dapr to panic

Reviewer: ```suggestion
We fixed a bug that caused Dapr to panic.
```

### Solution
 
We have implemented a fix in the component initialization sequence. Now, components that reference secrets from the Kubernetes secret store do not need the built-in Kubernetes secret store to be enabled anymore.

## Fixes panic in actor placement membership in Dapr sidecar

### Problem

When recovering from a failure in the connection to the Dapr placement service, the Dapr sidecar could have encountered a panic in some cases.

### Impact

The issue can impact all Dapr users on Dapr 1.7.0-1.9.4 using actors.

### Root cause

We identified a race condition in the actor placement service client that could have caused the Dapr to panic after recovering from a failure.

### Solution

We updated actor placement service to address the race condition and remove the cause for the panic.

## Fixes MQTT message acknowledgement for retained messages

### Problem

Retained MQTT messages should be only processed once, however, they were not being acknowledged and thus were resent indefinitely.

### Impact

> This issue impacts all users of the MQTT PubSub component on Dapr 1.9.0-1.9.4, who have enabled `retain` in the component's metadata.

Reviewer 2: ```suggestion
This issue impacts all users of the MQTT PubSub component on Dapr 1.9.0-1.9.4, where-in a subscriber using MQTT component receives a `retained` message.
```

### Solution
 
We have implemented a fix in the component initialization sequence. Now, components that reference secrets from the Kubernetes secret store do not need the built-in Kubernetes secret store to be enabled anymore.

## Fixes nil dereference crash in placement membership heartbeat loop in sidecar

### Problem

Dapr sidecar panics due to a nil dereference in the actor placement service.

### Impact

This issue impacts all users who are using actors.

### Root cause

> After closing the client stream in the actor placement service, the stream was being set to nil.

Reviewer: ```suggestion
We identified a race condition in the actor placement service client that could have caused the Dapr to panic after recovering from a failure.
```

### Solution

> We updated actor placement service to properly close the client stream and no longer set it to nil.

Reviewer: ```suggestion
We updated actor placement service to address the race condition and remove the cause for the panic.
```

### Solution

We updated actor placement service to address the race condition and remove the cause for the panic.

## Fixes MQTT message acknowledgement for retained messages

### Problem

> Retained mqtt messages should be only processed once, however, they were not being acknowledged and thus were resent indefinitely.

Reviewer: MQTT uppercase

### Impact

> This issue impacts all users of MQTT on Dapr 1.9.0-1.9.4.

Reviewer: ```suggestion
This issue impacts all users of the MQTT PubSub component on Dapr 1.9.0-1.9.4, who have enabled `retain` in the component's metadata.
```

### Root cause

MQTT retained messages were never being acknowledged and were therefore being resent indefinitely. 

### Solution

> We updated Dapr to acknowledge a retained message.

Reviewer: ```suggestion
We updated Dapr to acknowledge a retained message in the MQTT PubSub component.
```
